File: cpp/d3.cpp
#include "aoc.h"

int main() {
    ifstream file("puzzles/d3.txt");
    regex delim(R"((mul\(\d+,\d+\))|(don\'t\(\))|(do\(\)))");
    regex inner_delim(R"(^mul\((\d+),(\d+)\))");
    sregex_token_iterator end;
    string line;
    int ans1{}, ans2{};
    vector<string> all_search;
    while (getline(file, line)) {
        sregex_token_iterator it(
            line.begin(),
            line.end(),
            delim,
            0 // special value: return the parts *between* matches. 0 for the matches
        );
        for (; it != end; ++it) {
            all_search.push_back(it->str());
        }
    }
    bool on = true;
    for (auto x: all_search) {
        if (!(x=="don't()" || x=="do()")) {
            if (smatch match; regex_match(x, match, inner_delim)) {
                ans1 += stoi(match[1]) * stoi(match[2]);
            }
        }
        if (on && !(x=="don't()" || x=="do()")) {
            if (smatch match; regex_match(x, match, inner_delim)) {
                ans2 += stoi(match[1]) * stoi(match[2]);
            }
        }
        else if (x=="do()") {
            on = true;
        }
        else if (x=="don't()") {
            on = false;
        }
    }
    cout << ans1 << "  " << ans2 << endl;
}

File: js/d3.js
const fs = require("fs");

fs.readFile("puzzles/d3.txt", "utf-8", (lol, data) => {
    const lines = data.split(/\n|\r\n/);
    let ans1 = 0;
    let ans2 = 0;
    let enabled = true;
    lines.forEach((line) => {
        if (line == "") return;
        // Getting all mul(int, int), do() and don't()
        const res = line.match(/mul\(\d+,\d+\)|do\(\)|don\'t\(\)/g);
        res.forEach((mul) => {
            if (mul == "do()") {
                enabled = true;
            } else if (mul == "don't()") {
                enabled = false;
            } else {
                const nums = mul.match(/\d+/g);
                ans1 += nums[0] * nums[1];
                if (enabled) {
                    ans2 += nums[0] * nums[1];
                }
            }
        });
    });
    console.log("part1:", ans1, " part2:", ans2);
});
